linmod/models/utils.py

```diff
@@ -0,0 +1,221 @@
+
+    __call__ arguments
+    forecast:          A LazyFrame with the standard model output format.
+    geo_map:           A dictionary mapping "divisions" in the forecast to new units.
+                       Every existing division need not be mapped.
+
+    Returns a LazyFrame in the standard output format on the new geographical units.
+    """
+
+    @abstractmethod
+    def __call__(
+        self, forecast: pl.LazyFrame, geo_map: dict[str, str], **kwargs
+    ) -> pl.LazyFrame:
+        raise NotImplementedError()
+
+
+class InfectionWeightedAggregator(GeographicAggregator):
+    r"""
+    Geographically aggregate while accounting for unequal numbers of infected individuals in different states.
+
+    For convenience, this is broken down into per-state total human populations, and the proportions of those populations
+    which are infected. The per-state weight in the aggregation is the product of these terms.
+
+    The state-level populations and proportions infected are taken to be constant over the forecasting horizon.
+
+    __call__ keyword arguments
+    pop_size:          A pl.LazyFrame with the (possibly relative) population size of each forecasted division.
+                       Must contain all divisions which are in the `geo_map`, and two columns: "division" and "pop_size".
+                       Defaults to equal population sizes.
+    prop_infected      A pl.LazyFrame with the proportion of the populations infected (or something proportionate thereto),
+                       taken to be a constant over the modeling period. Must contain all divisions which are in the `geo_map`,
+                       and two columns: "division" and "prop_infected". Defaults to equal proportions infected.
+    """
+
+    def __call__(
+        self, forecast: pl.LazyFrame, geo_map: dict[str, str], **kwargs
+    ) -> pl.LazyFrame:
+        pop_size = kwargs.get(
+            "pop_size",
+            pl.DataFrame(
+                {
+                    "division": geo_map.keys(),
+                    "pop_size": [1.0] * len(geo_map.keys()),
+                }
+            ).lazy(),
+        )
+        prop_infected = kwargs.get(
+            "prop_infected",
+            pl.DataFrame(
+                {
+                    "division": geo_map.keys(),
+                    "prop_infected": [1.0] * len(geo_map.keys()),
+                }
+            ).lazy(),
+        )
+        weights = (
+            pop_size.join(prop_infected, on="division", how="inner")
+            .with_columns(
+                weight=(pl.col("pop_size") * pl.col("prop_infected"))
+            )
+            .select("division", "weight")
+        )
+
+        return (
+            forecast.filter(pl.col("division").is_in(geo_map.keys()))
+            .with_columns(
+                pl.col("division")
+                .replace_strict(geo_map)
+                .alias("new_division")
+            )
+            .join(weights, on="division", how="left")
```

Comment:
throw in a validation here? implicitly missing weight means not zero weight, but no prediction?

I'm just always nervous about joins now

Reply:
Yeah, we should probably check `geo_map`, `prop_infected`, and `pop_size` against each other in some way. Like, if you want to leave a state out, fine, don't put it in the `geo_map`, but if it's there it should be in both the dataframes.

Reply:
and then the test suite can reflect whatever failure modes we expect